crates/crypto/src/threshold_signing/indexed_signatures/verification.rs

```diff
@@ -0,0 +1,242 @@
+//! ECDSA signature verification for threshold signing.
+
+use secp256k1::{
+    ecdsa::{RecoverableSignature, RecoveryId},
+    Message, SECP256K1,
+};
+
+use super::{SignatureSet, ThresholdConfig, ThresholdSigningError};
+
+/// Verifies a set of ECDSA signatures against a threshold configuration.
+///
+/// # Arguments
+///
+/// * `config` - The threshold configuration containing authorized public keys
+/// * `signatures` - The set of indexed ECDSA signatures to verify
+/// * `message_hash` - The 32-byte message hash that was signed
+///
+/// # Verification Steps
+///
+/// 1. Check that the number of signatures meets the threshold
+/// 2. For each signature, verify that:
+///    - The signer index is within bounds
+///    - The ECDSA signature is valid for the corresponding public key
+///
+/// # Returns
+///
+/// * `Ok(())` if all signatures are valid and threshold is met
+/// * `Err(ThresholdSigningError)` otherwise
+pub fn verify_threshold_signatures(
+    config: &ThresholdConfig,
+    signatures: &SignatureSet,
+    message_hash: &[u8; 32],
+) -> Result<(), ThresholdSigningError> {
+    // Check threshold is met
+    if signatures.len() < config.threshold() as usize {
+        return Err(ThresholdSigningError::InsufficientSignatures {
+            provided: signatures.len(),
+            required: config.threshold() as usize,
+        });
+    }
+
+    // Create the message for verification
+    let message = Message::from_digest_slice(message_hash)
+        .map_err(|_| ThresholdSigningError::InvalidMessageHash)?;
+
+    // Verify each signature
+    for indexed_sig in signatures.signatures() {
+        // Check index is in bounds
+        let index = indexed_sig.index as usize;
+        if index >= config.keys().len() {
+            return Err(ThresholdSigningError::SignerIndexOutOfBounds {
+                index: indexed_sig.index,
+                max: config.keys().len(),
+            });
+        }
+
+        // Get the expected public key
+        let expected_pubkey = config.keys()[index].as_inner();
+
+        // Parse the recoverable signature
+        let recovery_id = RecoveryId::from_i32(indexed_sig.recovery_id() as i32)
+            .map_err(|_| ThresholdSigningError::InvalidSignatureFormat)?;
+
+        let recoverable_sig =
+            RecoverableSignature::from_compact(&indexed_sig.compact(), recovery_id)
+                .map_err(|_| ThresholdSigningError::InvalidSignatureFormat)?;
+
+        // Recover the public key from the signature
+        let recovered_pubkey = SECP256K1
+            .recover_ecdsa(&message, &recoverable_sig)
+            .map_err(|_| ThresholdSigningError::InvalidSignature {
+                index: indexed_sig.index,
+            })?;
+
+        // Verify the recovered key matches the expected key
+        if &recovered_pubkey != expected_pubkey {
+            return Err(ThresholdSigningError::InvalidSignature {
+                index: indexed_sig.index,
+            });
+        }
```

Comment:
What's the benefit, if any, of this verification (extract publey and match) on using the pubkey (fetched by index) an applying the sig verification algorithm?